pkg/importer/dataStream.go

```diff
@@ -266,6 +273,110 @@ func (d *DataStream) http() (io.ReadCloser, error) {
 	return countingReader, nil
 }
 
+type registryData struct {
+	dataDir string
+	file    *os.File
+}
+
+//_regData - global variable that stores temporary data of registy flow
+// to be deleted when teh flow ends
+var _regData = registryData{"", nil}
+
+type registryDataCloser struct {
+	io.Reader
+}
+
+func (registryDataCloser) Close() error {
+	_regData.file.Close()
+	os.RemoveAll(_regData.dataDir)
+	return nil
+}
+
+// RegistryDataCloser returns a ReadCloser with a Close method wrapping delettion
+// of temporary data of registry flow
+func RegistryDataCloser(r io.Reader) io.ReadCloser {
+	return registryDataCloser{r}
+}
+
+const (
+	//ContainerDiskImageDir - Expected disk image location in contianer image as described in
+	//https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md
+	ContainerDiskImageDir = "disk"
+	//TempContainerDiskDir - Temporary location for pulled containerImage
+	TempContainerDiskDir = "tmp"
+)
+
+//This import source downloads specified container image from registry location
+//Then it extracts the image to a temporary location and expects an image file to be located under /disk directory
+//If such exists it creates a Reader on it and returns it for further processing
+func (d *DataStream) registry() (io.ReadCloser, error) {
+	//generate random name for temporary directory
+	_regData.dataDir = filepath.Join(d.DataDir, TempContainerDiskDir)
+
+	imageDir := filepath.Join(_regData.dataDir, ContainerDiskImageDir)
+
+	//create temporary directory if note exists to which all the data will be extracted
+	if _, err := os.Stat(_regData.dataDir); os.IsNotExist(err) {
+		err := os.Mkdir(_regData.dataDir, os.ModeDir|os.ModePerm)
+		if err != nil {
+			return nil, errors.Wrapf(err, fmt.Sprintf("Failed to create tempdirectory %s", _regData.dataDir))
+		}
+	}
+
+	//copy image from registry to the temporary location
+	glog.V(1).Infof("using skopeo to copy from registry")
+	err := image.CopyDirFromRegistryImage(d.Endpoint, _regData.dataDir, ContainerDiskImageDir, d.AccessKey, d.SecKey)
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Failed ro read from registry"))
+	}
+
+	//Search for file in /disk directory - if not found - failure
+	imageFile, err := getImageFileName(imageDir)
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Cannot locate image file"))
+	}
+
+	// 3. If found - Create a reader that will read this file and attach it to the dataStream
+	_regData.file, err = os.Open(filepath.Join(imageDir, imageFile))
+	if err != nil {
+		os.RemoveAll(_regData.dataDir)
+		return nil, errors.Wrapf(err, fmt.Sprintf("Fail to create data stream from image file"))
+	}
+
+	glog.V(1).Infof("VM disk image filename is %s", imageFile)
+
+	return RegistryDataCloser(bufio.NewReader(_regData.file)), nil
```

Comment:
this seems really inefficient and could cause issues with PVC capacity.  Basically, the PVC has to be at least twice the size of the image because you are doing a copy.

Reply:
First of all, I think we should discuss how is it better to solve temporary space issue required by all import flows.

Regarding the specific flow:
The VM disk image image we import can be in several formats:
- raw - (no redundancy see below)
- qcow2 - agree 1 redundant copy  
- qcow archived - (no redundancy see below)

- In case of raw - there is no redundant copy  - we extract the disk image file from container image and then copy it to /disk/disk.img - 
**tmpPVC** space = sizeof(containerImage)+sizeof(/disk directory in the image ~  sizeof(VM disk image file) in typical registry Image) 
**PVC space** = sizeof (raw VM disk image)

- In case of qcow2 archived - there is no redundant copy - we extract archived qcow2 file from container image and then in StreamDataToFile we unarchive it and then convert from qcow2 to raw.
**tmpPVC space** = sizeof(containerImage) + sizeof(~ archived VM disk image file)  + sizeof (unarchived VM disk image file)
**PVC space** = sizeof (raw VM disk image file)

In case of unarchive qcow2 - I agree that one copy is redundant (I am handling it) - currently 
**tmpPVC space** = sizeof(containerImage) + 2*sizeof(~  VM disk image file in qcow2 format)
**PVC space** = sizeof (raw VM disk image file)

After the fix space consumption is:
**tmpPVC space** = sizeof(containerImage) + sizeof(~  VM disk image file in qcow2 format)
**PVC space** = sizeof (raw VM disk image file)

Reply:
I see redundancy in all cases.  Here's how I see it, ignoring downloading of the container layers which obliviously adds its own overhead.

Raw:

0.  Image extracted from image to `/data/disk/myimage.raw`
1.  Image COPIED from `/data/disk/myimage.raw` to `/data/disk.img`
3.  `/data/disk` directory deleted.

At step 1, there are two copies of the image on the PVC.

qcow2:

0.  Image extracted from image to `/data/disk/myimage.qcow2`
1.  Image copied to `/data/<temp dir>/myimage.qcow2`
2.  Conversion to raw creates `/data/disk.img`
3.  `/data/<temp dir>` deleted
4.  `/data/disk` deleted

At step 2, there are two copies of the qcow2 image on the pvc

qcow2 archived is similar to above except replace step 0 with:

0.  Image extracted from image to `/data/disk/myimage.qcow2.tar.gz`

Reply:
optimized in case of qcow2 file - now it is not copied to additional tmp location, but converted to raw from the location registry flow extracted it to